service/header/exchange.go

```diff
@@ -0,0 +1,194 @@
+
+import (
+	"context"
+	"fmt"
+
+	pb "github.com/celestiaorg/celestia-node/service/header/pb"
+	"github.com/celestiaorg/go-libp2p-messenger/serde"
+
+	"github.com/libp2p/go-libp2p-core/host"
+	"github.com/libp2p/go-libp2p-core/network"
+	"github.com/libp2p/go-libp2p-core/peer"
+	"github.com/libp2p/go-libp2p-core/protocol"
+)
+
+var headerExchangeProtocolID = protocol.ID("header-exchange")
+
+// exchange enables sending outbound ExtendedHeaderRequests as well as
+// handling inbound ExtendedHeaderRequests.
+type exchange struct {
+	host host.Host
+	// TODO @renaynay: post-Devnet, we need to remove reliance of Exchange on one bootstrap peer
+	// Ref https://github.com/celestiaorg/celestia-node/issues/172#issuecomment-964306823.
+	peer  *peer.AddrInfo
+	store Store
+}
+
+func newExchange(host host.Host, peer *peer.AddrInfo, store Store) *exchange {
+	ex := &exchange{
+		host:  host,
+		peer:  peer,
+		store: store,
+	}
+	host.SetStreamHandler(headerExchangeProtocolID, ex.requestHandler)
+	return ex
+}
+
+// requestHandler handles inbound ExtendedHeaderRequests.
+func (e *exchange) requestHandler(stream network.Stream) {
+	// unmarshal request
+	pbreq := new(pb.ExtendedHeaderRequest)
+	_, err := serde.Read(stream, pbreq)
+	if err != nil {
+		log.Errorw("reading header request from stream", "err", err.Error())
+		//nolint:errcheck
+		stream.Reset()
+		return
+	}
+	request := &ExtendedHeaderRequest{
+		Origin: pbreq.Origin,
+		Amount: pbreq.Amount,
+	}
+	// route depending on amount of headers requested
+	if request.Amount > 1 {
+		e.handleMultipleHeaderRequest(request, stream)
+	} else {
+		e.handleSingleHeaderRequest(request.Origin, stream) // TODO @renaynay: should we parallelise this?
+	}
+}
+
+// handleSingleHeaderRequest handles an ExtendedHeaderRequest for a single header and writes
+// the header to the given stream.
+func (e *exchange) handleSingleHeaderRequest(origin uint64, stream network.Stream) {
+	var (
+		header *ExtendedHeader
+		err    error
+	)
+	if origin == uint64(0) {
+		header, err = e.store.Head()
+	} else {
+		header, err = e.store.GetByHeight(context.Background(), origin)
+	}
+	if err != nil {
+		log.Errorw("getting header by height", "height", origin, "err", err.Error())
+		//nolint:errcheck
+		stream.Reset()
+		return
+	}
+	resp, err := ExtendedHeaderToProto(header)
+	if err != nil {
+		log.Errorw("marshaling header to proto", "height", origin, "err", err.Error())
+		//nolint:errcheck
+		stream.Reset()
+		return
+	}
+	_, err = serde.Write(stream, resp)
+	if err != nil {
+		log.Errorw("writing header to stream", "height", origin, "err", err.Error())
+	}
+}
+
+// handleMultipleHeaderRequest handles responding to an ExtendedHeaderRequest for
+// multiple headers.
+func (e *exchange) handleMultipleHeaderRequest(request *ExtendedHeaderRequest, stream network.Stream) {
+	height := request.Origin
+	amount := uint64(0)
+	for amount < request.Amount {
+		e.handleSingleHeaderRequest(height, stream)
+		height++
+		amount++
+	}
```

Comment:
Once you done writing, you must close the stream to free up allocated resources.